cpp/src/arrow/compute/kernel.h

```diff
@@ -141,6 +141,14 @@ ARROW_EXPORT std::shared_ptr<TypeMatcher> FixedSizeBinaryLike();
 // Type)
 ARROW_EXPORT std::shared_ptr<TypeMatcher> Primitive();
 
+/// \brief Match run-end encoded types that encode specific plain data types
+///
+/// @param[in] run_ends_type_matcher a matcher that is applied to the run_ends field
+/// @param[in] encoded_type_matcher a matcher that is applied to the values field
+ARROW_EXPORT std::shared_ptr<TypeMatcher> RunEndEncoded(
+    std::shared_ptr<TypeMatcher> run_ends_type_matcher,
```

Comment:
Is it generally useful to pass a run ends matcher? The implementation will probably be the same, just templated by run ends type.

Reply:
Not really. Instances of this are passing `Primitive()`. I fixed `Matches` to match the integer types more strictly now. And as you said, kernel implementations will template on them on the three possible types.